flink/v1.17/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceWithWatermarkExtractor.java

```diff
@@ -0,0 +1,375 @@
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
+import org.apache.flink.streaming.api.functions.windowing.AllWindowFunction;
+import org.apache.flink.streaming.api.operators.collect.CollectResultIterator;
+import org.apache.flink.streaming.api.operators.collect.CollectSinkOperator;
+import org.apache.flink.streaming.api.operators.collect.CollectSinkOperatorFactory;
+import org.apache.flink.streaming.api.operators.collect.CollectStreamSink;
+import org.apache.flink.streaming.api.windowing.assigners.TumblingEventTimeWindows;
+import org.apache.flink.streaming.api.windowing.time.Time;
+import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.test.util.MiniClusterWithClientResource;
+import org.apache.flink.util.Collector;
+import org.apache.iceberg.FileFormat;
+import org.apache.iceberg.data.GenericAppenderHelper;
+import org.apache.iceberg.data.GenericRecord;
+import org.apache.iceberg.data.RandomGenericData;
+import org.apache.iceberg.data.Record;
+import org.apache.iceberg.flink.HadoopTableResource;
+import org.apache.iceberg.flink.RowDataConverter;
+import org.apache.iceberg.flink.TestFixtures;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.types.Comparators;
+import org.apache.iceberg.util.StructLikeWrapper;
+import org.awaitility.Awaitility;
+import org.junit.Assert;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public class TestIcebergSourceWithWatermarkExtractor implements Serializable {
+  private static final InMemoryReporter reporter = InMemoryReporter.createWithRetainedMetrics();
+  private static final int PARALLELISM = 4;
+  private static final String SOURCE_NAME = "IcebergSource";
+  private static final int RECORD_NUM_FOR_2_SPLITS = 200;
+
+  @ClassRule public static final TemporaryFolder TEMPORARY_FOLDER = new TemporaryFolder();
+
+  @Rule
+  public final MiniClusterWithClientResource miniClusterResource =
+      new MiniClusterWithClientResource(
+          new MiniClusterResourceConfiguration.Builder()
+              .setNumberTaskManagers(1)
+              .setNumberSlotsPerTaskManager(PARALLELISM)
+              .setRpcServiceSharing(RpcServiceSharing.DEDICATED)
+              .setConfiguration(reporter.addToConfiguration(new Configuration()))
+              .withHaLeadershipControl()
+              .build());
+
+  @Rule
+  public final HadoopTableResource sourceTableResource =
+      new HadoopTableResource(
+          TEMPORARY_FOLDER, TestFixtures.DATABASE, TestFixtures.TABLE, TestFixtures.TS_SCHEMA);
+
+  @Test
+  public void testWindowing() throws Exception {
```

Comment:
I am not very clear about our test strategy to confirm watermark alignment. We can probably borrow some ideas from Flink test code on watermark alignment

Reply:
This is how the watermark alignment is tested in the Flink code - basically the metrics are tested:
```
      // Get the drift metric, wait for it to be created and reach the expected state
      // (100 min - 20 min - 0 min)
      // Also this validates that the WatermarkAlignment is working
      Awaitility.await()
          .atMost(120, TimeUnit.SECONDS)
          .until(() -> findAlignmentDriftMetric(jobClient.getJobID(), 4800000L).isPresent());
      Gauge<Long> drift = findAlignmentDriftMetric(jobClient.getJobID(), 4800000L).get();
[..]
      // Get the drift metric, wait for it to be created and reach the expected state (100 min - 20
      // min - 15 min)
      Awaitility.await().atMost(120, TimeUnit.SECONDS).until(() -> drift.getValue() == 3900000L);
[..]
      // Wait for the new drift to decrease below the allowed drift to signal the normal state
      Awaitility.await().atMost(120, TimeUnit.SECONDS).until(() -> drift.getValue() < 1200000L);
```

Other than the metrics, we do an e2e testing by making sure that by adding new records we unblock the readers, and new records are emitted:
```
      // Add some new records which should unblock the throttled reader
      batch = ImmutableList.of(generateRecord(110, "File-5-110"), generateRecord(111, "File-5-111"));
      dataAppender.appendToTable(batch);
      // We should get all the records at this point
      waitForRecords(resultIterator, 6);
```

Reply:
> * d lateness to 20, make sure 3 splits with target split size
> * slow down the processing of file 1 and 2 with some sleep. make sure the watermark advanced to 50 before processing file 3.

The `testThrottling` does something similar. We use the `findAlignmentDriftMetric` to validate that the watermarks is as we expect it.

> * test A: use sequence number ordered assigner and allowed lateness. the records from file 3 should have been dropped and not in the final output

The watermarks are generally handled on task manager level. If the 3rd split (with watermark 10) is assigned to the task manager which read the 1st split (watermark 1), then they will not be late records. If it is assigned to the task manager which read the 2nd split (watermark 50), they will be late records. There is no easy way to ensure the scheduling happens on the task manager which we prefer.

> * test B: enable watermark alignment. records from all 3 files should be kept in the final output

For testing the ordering I will add a separate test class.

Reply:
Added `TestWatermarkBasedSplitAssigner` which also test the ordering based on the watermark comparator